lib/cheers_leader.rb

```diff
@@ -1,21 +1,31 @@
 class CheerLeaderList
+  attr_reader :list
+  
   def initialize(list)
+    raise ArgumentError if !list.is_a?(Array) || list.any? {|x| x.any? { |i| i.nil? } }
     @list = list
+    sort_by_rank # Why this not WORKING ?? 
+  end
+
+  def sort_by_rank(asc = true)
+    asc ? list.sort_by {|x| x[:rank] } : list.sort_by {|x| -x[:rank] }
+  end
+
+  def sort_by_name(asc = true)
+    #  WHY -x[1] NOT WORKING here ? 
+    asc ? list.sort_by {|x| x[:name] } : list.sort {|x, y| y[:name] <=> x[:name] }
+  end
+
+  def sort_by_sent(asc = true)
+    asc ? list.sort_by {|x| x[2] } : list.sort_by {|x| -x[2] }
+  end
+
+  def sort_by_received(asc = true)
+    asc ? list.sort_by {|x| x[3] } : list.sort_by {|x| -x[3] }
   end
-attr_reader :list
 
-  def sort_by(sort_by = 'rank', asc = true)
-    return [] if list.to_a.empty?
-    case sort_by
-      when 'name'
-        asc ? list.sort_by {|x| x[1] } : list.sort {|x, y| y[1] <=> x[1] }
-      when 'sent'
-        asc ? list.sort_by {|x| x[2] } : list.sort_by {|x| -x[2] }  
-      when 'received'
-        asc ? list.sort_by {|x| x[3] } : list.sort_by {|x| -x[3] }  
-      else
-        asc ? list.sort_by {|x| x[0] } : list.sort_by {|x| -x[0] }  
-    end
+  def search(text)
+    list.select {|x| x[1].match (text) }
```

Comment:
I'm pretty confused whether you want to use `x[1]` or `x[:rank]` or `x.rank` :)

Reply:
Thanks, this missing